halo2_proofs/src/plonk/circuit.rs

```diff
@@ -1545,6 +1685,472 @@ impl<F: Field> Gate<F> {
+    k: u32,
+    n: usize,
+    unusable_rows_start: usize,
+    circuit: &'a ConcreteCircuit,
+    config: &'a ConcreteCircuit::Config,
+    cs: &'a ConstraintSystem<F>,
+    instances: &'a [&'a [F]],
+    next_phase: u8,
+}
+
+impl<'a, F: Field, ConcreteCircuit: Circuit<F>> WitnessCalculator<'a, F, ConcreteCircuit> {
+    /// Create a new WitnessCalculator
+    pub fn new(
+        k: u32,
+        circuit: &'a ConcreteCircuit,
+        config: &'a ConcreteCircuit::Config,
+        cs: &'a ConstraintSystem<F>,
+        instances: &'a [&'a [F]],
+    ) -> Self {
+        let n = 2usize.pow(k);
+        let unusable_rows_start = n - (cs.blinding_factors() + 1);
+        Self {
+            k,
+            n,
+            unusable_rows_start,
+            circuit,
+            config,
+            cs,
+            instances,
+            next_phase: 0,
+        }
+    }
+
+    /// Calculate witness at phase
+    pub fn calc(
+        &mut self,
+        phase: u8,
+        challenges: &HashMap<usize, F>,
+        // ) -> Result<Vec<Option<Polynomial<Assigned<F>, LagrangeCoeff>>>, Error> {
+    ) -> Result<Vec<Option<Vec<Assigned<F>>>>, Error> {
+        if phase != self.next_phase {
+            return Err(Error::Other(format!(
+                "Expected phase {}, got {}",
+                self.next_phase, phase
+            )));
+        }
+        let current_phase = match phase {
+            0 => FirstPhase.to_sealed(),
+            1 => SecondPhase.to_sealed(),
+            2 => ThirdPhase.to_sealed(),
+            _ => unreachable!("only phase [0,2] supported"),
+        };
+        let mut witness = WitnessCollection {
+            k: self.k,
+            current_phase,
+            advice: vec![vec![Assigned::Zero; self.n]; self.cs.num_advice_columns],
+            instances: self.instances,
+            challenges,
+            // The prover will not be allowed to assign values to advice
+            // cells that exist within inactive rows, which include some
+            // number of blinding factors and an extra row for use in the
+            // permutation argument.
+            usable_rows: ..self.unusable_rows_start,
+            _marker: std::marker::PhantomData,
+        };
+
+        // Synthesize the circuit to obtain the witness and other information.
+        ConcreteCircuit::FloorPlanner::synthesize(
+            &mut witness,
+            self.circuit,
+            self.config.clone(),
+            self.cs.constants.clone(),
+        )
+        .expect("todo");
+
+        let column_indices = self
+            .cs
+            .advice_column_phase
+            .iter()
+            .enumerate()
+            .filter_map(|(column_index, phase)| {
+                if current_phase == *phase {
+                    Some(column_index)
+                } else {
+                    None
+                }
+            })
+            .collect::<BTreeSet<_>>();
```

Comment:
This can be done via `par_iter` too by default.

Reply:
I think this function is just too small for parallelization, it's just an iteration over max 1000 items, and for each one there's only an integer comparison.
I think parallelizing this would make it slower due to threading and result merging overhead.

Reply:
Totally right!